rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/RemoteFetchProcessor.java

```diff
@@ -0,0 +1,207 @@
+    while (!processQueue.isEmpty()) {
+      FetchRequest request = processQueue.peek();
+      if (currentTime - request.getRequestedTime() < this.requestProcessDelay) {
+        break;
+      }
+
+      if (!rangeMap.containsKey(request.getRemotePath())) {
+        RangeSet<Long> rangeSet = TreeRangeSet.create();
+        rangeMap.putIfAbsent(request.getRemotePath(), rangeSet);
+        fetchRequestMap.putIfAbsent(request.getRemotePath(), new FileMetadataRequest(request.getRemotePath(),
+            request.getFileSize(), request.getLastModified()));
+      }
+      rangeMap.get(request.getRemotePath()).add(Range.open(request.getOffset(),
+          request.getOffset() + request.getLength()));
+      processQueue.remove();
+    }
+
+    processRemoteFetchRequest();
+
+    // After every iteration we are clearing the maps
+    rangeMap.clear();
+    fetchRequestMap.clear();
+  }
+
+  @Override
+  protected Scheduler scheduler()
+  {
+    return Scheduler.newFixedDelaySchedule(processThreadInitalDelay, processThreadInterval, TimeUnit.MILLISECONDS);
+  }
+
+  private void processRemoteFetchRequest() throws IOException, InterruptedException, ExecutionException
+  {
+    final List<FileDownloadRequestChain> readRequestChainList = new ArrayList<FileDownloadRequestChain>();
+
+    for (Iterator<Map.Entry<String, RangeSet<Long>>> it = rangeMap.entrySet().iterator(); it.hasNext(); ) {
+      Map.Entry<String, RangeSet<Long>> entry = it.next();
+      Path path = new Path(entry.getKey());
+      FileSystem fs = path.getFileSystem(conf);
+      fs.initialize(path.toUri(), conf);
+      FSDataInputStream inputStream = fs.open(path);
+      String localPath = CacheConfig.getLocalPath(entry.getKey(), conf);
+      log.info("Processing Request for File : " + path.toString() + " LocalFile : " + localPath);
+      FileMetadataRequest fileRequestMetadata = fetchRequestMap.get(entry.getKey());
+      ByteBuffer directWriteBuffer = bufferPool.getBuffer(diskReadBufferSize);
+
+      FileDownloadRequestChain requestChain = new FileDownloadRequestChain(fs, localPath,
+          directWriteBuffer, conf, fileRequestMetadata.remotePath, fileRequestMetadata.fileSize,
+          fileRequestMetadata.lastModified);
+
+      for (Range<Long> range : entry.getValue().asRanges()) {
+        log.info("Adding request for File : " + entry.getKey() + " Start : "
+            + range.lowerEndpoint() + " End : " + range.upperEndpoint());
+        ReadRequest request = new ReadRequest(range.lowerEndpoint(), range.upperEndpoint(),
+            range.lowerEndpoint(), range.upperEndpoint(), null, 0, fetchRequestMap.get(entry.getKey()).fileSize);
+        requestChain.addReadRequest(request);
+      }
+      it.remove();
+
+      log.info("Request added for file: " + requestChain.getRemotePath() + " Number of Requests : " +
+              requestChain.getReadRequests().size());
+      readRequestChainList.add(requestChain);
+    }
+
+    processDownloadRequests(readRequestChainList);
```

Comment:
Is the scheduler blocked until all the tasks are done ? Since the code calls `futures.get` ? Just For My Information. Also I am assuming you have a SchedulerThread for Range -> RemoteFetch to improve chances of batching. You could have directly put into FetchRequest but then there will be no batching.